inginious/frontend/pages/course_admin/student_list.py

```diff
@@ -192,35 +198,72 @@ def post_audiences(self, course, data, active_tab, msg, error):
-                else:
-                    audiences = json.loads(data["audiences"])
-
-                for index, new_audience in enumerate(audiences):
-                    # In case of file upload, no id specified
-                    new_audience['_id'] = new_audience['_id'] if '_id' in new_audience else 'None'
-
-                    # Update the audience
-                    audience, errors = self.update_audience(course, new_audience['_id'], new_audience)
-
-                    # If file upload was done, get the default audience id
-                    audienceid = audience['_id']
-                    errored_students += errors
+            if "audiencefile" in data and 'upload_audiences_creation' in data:
+                # get the Werkzeug datastructures.FileStorage object.
+                # The stream of this object is the stream body of the uploaded file.
+                # Furthermore, FileStorage.stream seems to inherit ʻio.BufferedIOBase`, so this stream should be boiled.
+                # As reader return an iterator and that we iterate twice, it is faster to cast into a list.
+                csv_data = list(csv.reader(io.TextIOWrapper(data["audiencefile"], encoding='utf-8')))
+                auth_method_ids = self.user_manager.get_auth_methods().keys()
+                students_per_audience = {}
+                tutors_per_audience = {}
+                course_students = []
+                course_tutors = []
+                audiences = []
+                for line in csv_data:
+                    if len(line) != 4:
+                        msg["audiences"] = _("File wrongly formatted.")
+                        error["audiences"] = True
+                if "audiences" not in error or not error["audiences"]:
+                    stud_list, aud_li, oth_stu, u_info = self.get_user_lists(course)
+                    courseid = course.get_id()
+                    self.database.audiences.delete_many({"courseid": courseid})
+                    for user_id, field, role, description in csv_data:
+                        user_id = user_id.strip()
+                        field = field.strip()
+                        role = role.strip()
+                        description = description.strip()
+                        if field not in ["username", "email", "realname"] + list(auth_method_ids) \
+                                or role not in ["student", "tutor"]:
+                            msg["audiences"] = _("Some fields were not recognized:" + field)
+                            error["audiences"] = True
+                            continue
+                        if field != "username":
+                            user = self.database.users.find_one({field: user_id})
+                            if user is not None:
+                                user_id = user["username"]
+                            else:
+                                msg["audiences"] = _("Some users were not found: " + user_id)
```

Comment:
You need to extract the dynamic part from the gettext string. Do not use plural form if your code only and always returns one user at a time.